compiler/z/codegen/S390Peephole.cpp

```diff
@@ -1349,26 +1349,66 @@ TR_S390PostRAPeephole::ConditionalBranchReduction(TR::InstOpCode::Mnemonic branc
 
 
 /**
- * Catch the pattern where an CR/BRC can be converted
- *    CR R1,R2
- *    BRC  Mask, Lable
+ * Catch the pattern where an CLR/BRC can be converted
+ *    CLR R1, R2
+ *    BRC Mask, Lable
  * Can be replaced with
- *    CRJ  R1,R2,Lable
+ *    CLRJ R1, R2, Lable, Mask
  */
 bool
 TR_S390PostRAPeephole::CompareAndBranchReduction()
    {
-   bool performed = false;
+   if (!TR::Compiler->target.cpu.getSupportsArch(TR::CPU::z10))
+      return false;
 
-   TR::Instruction *current = _cursor;
-   TR::Instruction *next = _cursor->getNext();
-   TR::InstOpCode::Mnemonic curOpCode = current->getOpCodeValue();
-   TR::InstOpCode::Mnemonic nextOpCode = next->getOpCodeValue();
-   if ((curOpCode == TR::InstOpCode::CR || curOpCode == TR::InstOpCode::CGR || curOpCode == TR::InstOpCode::CGFR)
-      && (nextOpCode == TR::InstOpCode::BRC || nextOpCode == TR::InstOpCode::BRCL))
-      {
-      printf("Finding CR + BRC\n");
-      printf("method=%s\n", comp()->signature());
+   bool performed = false;
+   TR::Compilation *comp = _cg->comp();
+   TR::S390RRInstruction *clrInstruction = (TR::S390RRInstruction*)_cursor;
+   TR::S390BranchInstruction *brcInstruction = (TR::S390BranchInstruction*)_cursor->getNext();
+   TR::Instruction *clrjInstruction = NULL;
+
+   if (clrInstruction->getOpCodeValue() == TR::InstOpCode::CLR
+       && brcInstruction->getOpCodeValue() == TR::InstOpCode::BRC)
+      {
+      /* if branch target is a snippet we'll go ahead with the
+       * reduction. Otherwise, we look at next 5 (or until TR::BBEnd
+       * depending on whichever comes first) to check if any
+       * instruction reads or sets the condition code.
+       * We'll perform the reduction if an instruction sets
+       * the condition code or, if none of the instructions read the
+       * condition code set by clrInstruction
+       */
```

Comment:
This is not safe. If the target of the branch is a snippet then you can assume the taken side of the branch is safe for reduction, however you still need to examine the fall-through side of the branch. i.e. you need to do the following:

For some peephole window of instructions (you picked 5 here which is reasonable):

- Examine the taken side of the branch to validate that no instruction reads the CC before it is set again
- Examine the fall-through side of the branch to validate that no instruction reads the CC before it is set again

If the target is a snippet you can skip the first check, however we still need to perform the second check.